src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -167,41 +183,155 @@ internal int EndRead(IAsyncResult asyncResult)
+                }
+                finally
+                {
+                    _nestedWrite = 0;
+                }
             }
+        }
 
-            // No "artificial" timeouts implemented so far, InnerStream controls timeout.
-            lazyResult.InternalWaitForCompletion();
+        private async Task WaitForWriteIOSlot(Task lockTask, byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            await lockTask.ConfigureAwait(false);
+            await WriteSingleChunk(buffer, offset, count, cancellationToken).ConfigureAwait(false);
+        }
+
+        internal Task WriteSingleChunk(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            // Request a write IO slot.
+            Task ioSlot = _sslState.CheckEnqueueWriteAsync();
+            if (!ioSlot.IsCompletedSuccessfully)
+            {
+                // Operation is async and has been queued, return.
+                return WaitForWriteIOSlot(ioSlot, buffer, offset, count, cancellationToken);
+            }
+
+            byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(count + FrameOverhead);
+            byte[] outBuffer = rentedBuffer;
+
+            SecurityStatusPal status = _sslState.EncryptData(buffer, offset, count, ref outBuffer, out int encryptedBytes);
+
+            if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+            {
+                // Re-handshake status is not supported.
+                ProtocolToken message = new ProtocolToken(null, status);
+                ArrayPool<byte>.Shared.Return(rentedBuffer);
+                throw new IOException(SR.net_io_encrypt, message.GetException());
+            }
+
+            Task t = _sslState.InnerStream.WriteAsync(outBuffer, 0, encryptedBytes, cancellationToken);
+            if (t.IsCompleted)
+            {
+                ArrayPool<byte>.Shared.Return(rentedBuffer);
+                _sslState.FinishWrite();
+                return t;
+            }
+            else
+            {
+                return CompleteAsync(t, rentedBuffer);
+            }
+
+            async Task CompleteAsync(Task writeTask, byte[] bufferToReturn)
+            {
+                try
+                {
+                    await writeTask.ConfigureAwait(false);
+                }
+                finally
+                {
+                    ArrayPool<byte>.Shared.Return(bufferToReturn);
+                    _sslState.FinishWrite();
+                }
+            }
+        }
 
-            if (lazyResult.Result is Exception e)
+        internal async Task WriteAsyncChunked(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            try
             {
+                do
+                {
+                    int chunkBytes = Math.Min(count, _sslState.MaxDataSize);
+                    await WriteSingleChunk(buffer, offset, chunkBytes, cancellationToken).ConfigureAwait(false);
```

Comment:
How likely is it that we won't be able to write all `count` bytes in one go?  There's overhead happening per call to WriteSingleChunk, e.g. renting and returning an array to the pool, checking for a write slot, etc... if it'll be common to split writes, we should look at ammortizing those costs.

Reply:
When `count` > 16kB; I imagine most network write calls will be smaller than 16kB chunks as time to first byte is an important metric

Reply:
Yeah, very rarely do you see > 16k, even if you read from the disk SSD's come in 4k, 8k and 16k pages. It's a total of 16k of actual plain text data. Also as a for instance the current buffer size in pipelines is 4k (minus a little for reasons I won't get into here) to align with memory pages so it won't ever try to write > 4k at a time.

So I would say 95% of times you are less than the 16k boundary.